src/uct/ib/base/ib_iface.c

```diff
@@ -1173,6 +1174,118 @@ static void uct_ib_iface_set_path_mtu(uct_ib_iface_t *iface,
     }
 }
 
+const char*
+uct_ib_iface_sl_mask_str(uint16_t sl_mask, char *str, size_t str_length)
+{
+    size_t offset = 0;
+    uint8_t sl;
+
+    if (sl_mask == 0) {
+        ucs_strncpy_zero(str, "<none>", str_length);
+        goto out;
+    }
+
+    ucs_for_each_bit(sl, sl_mask) {
+        ucs_assert(sl < 16);
+
+        if (offset == 0) {
+            ucs_snprintf_safe(str, str_length, "%u", sl);
+        } else {
+            ucs_snprintf_safe(UCS_PTR_BYTE_OFFSET(str, offset),
+                              str_length - offset, ", %u", sl);
+            offset += 2 /* for comma and whitespace */;
+        }
+
+        offset += (sl < 10) ? 1 : 2;
+    }
+
+out:
+    return str;
+}
+
+
+ucs_status_t
+uct_ib_iface_select_sl(const uct_ib_iface_config_t *ib_config,
+                       ucs_ternary_value_t ar_enable, uint16_t ooo_sl_mask,
+                       const char *dev_name, uint8_t port_num, uint8_t *sl_p)
+{
+    char sl_str[64] = "";
+    uint8_t sl;
+
+    /* Select the default  */
+    ucs_assert((ib_config->sl <= UINT8_MAX) || (ib_config->sl == UCS_ULUNITS_AUTO));
+    sl = (ib_config->sl == UCS_ULUNITS_AUTO) ? 0 : (uint8_t)ib_config->sl;
+
+    if (ar_enable == UCS_NO) {
+        goto out;
+    }
+
+    if (ib_config->sl == UCS_ULUNITS_AUTO) {
+        if (ooo_sl_mask != 0) {    
+            /* Select the first available SL with AR support, since a specific
+             * SL was not requested by user */
+           sl = ucs_ffs64_safe(ooo_sl_mask);
+           ucs_assert(sl != 64);
+        } else {
+            /* No SLs support AR */
+            if (ar_enable == UCS_YES) {
+                ucs_error("AR support is requested, but no SLs support AR on"
+                          " %s:%u", dev_name, port_num);
+                return UCS_ERR_UNSUPPORTED;
+            } else {
+                ucs_diag("AR support is not strongly required and no SLs"
+                         " support AR on %s:%u", dev_name, port_num);
+                goto out;
+            }
+        }
+    } else if (!(ooo_sl_mask & UCS_BIT(ib_config->sl))) {
+        if (ar_enable == UCS_YES) {
+            /* User requires AR, but selected SL doen'st support AR,
+             * fail selection */
+            ucs_error("AR support is requested, SL=%u was choosen,"
+                      " but this SL doesn't support AR,"
+                      " { %s } - other SLs that support AR on %s:%u", sl,
+                      uct_ib_iface_sl_mask_str(ooo_sl_mask, sl_str,
+                                               sizeof(sl_str)),
+                      dev_name, port_num);
+            return UCS_ERR_UNSUPPORTED;
+        }
+
+        /* AR support is not strongly required by user and selected SL
+         * doesn't support AR */
+        ucs_assert(ar_enable == UCS_TRY);
+        ucs_diag("SL=%u is choosen, but only { %s } SLs support AR on %s:%u",
```

Comment:
I'd use debug instead, it will be a typical situation flooding diag level

Reply:
done